pkg/operator/watcher/filewatcher.go

```diff
@@ -0,0 +1,112 @@
+package watcher
+
+import (
+	"fmt"
+	"io/ioutil"
+	"os"
+	"sync"
+
+	"gopkg.in/fsnotify.v1"
+
+	logf "github.com/openshift/cluster-ingress-operator/pkg/log"
+)
+
+var log = logf.Logger.WithName("filewatcher")
+
+// FileWatcher watches a file for changes.
+type FileWatcher struct {
+	sync.Mutex
+	fileName string
+	watcher  *fsnotify.Watcher
+}
+
+// New returns a new FileWatcher watching the given file.
+func New(file string) (*FileWatcher, error) {
+	fw := &FileWatcher{
+		fileName: file,
+	}
+
+	// Initial read of file.
+	if err := fw.ReadFile(); err != nil {
+		return nil, err
+	}
+
+	if watcher, err := fsnotify.NewWatcher(); err != nil {
+		return nil, err
+	} else {
+		fw.watcher = watcher
+	}
+
+	return fw, nil
+}
+
+// Start starts the FileWatcher.
+func (fw *FileWatcher) Start(stopCh <-chan struct{}) error {
+	if err := fw.watcher.Add(fw.fileName); err != nil {
+		return err
+	}
+
+	go fw.Watch()
+	log.Info("starting file watcher")
+	// Block until the stop channel is closed.
+	<-stopCh
+
+	return fw.watcher.Close()
+}
+
+// Watch reads events from the watcher's channel and reacts to changes.
+func (fw *FileWatcher) Watch() {
+	for {
+		select {
+		case event, ok := <-fw.watcher.Events:
+			if !ok {
+				log.Info("file watch events channel closed")
+				return
+			}
+			fw.handleEvent(event)
+
+		case err, ok := <-fw.watcher.Errors:
+			if !ok {
+				log.Info("file watch error channel closed")
+				return
+			}
+			log.Error(err, "file watch error")
+		}
+	}
+}
+
+// ReadFile reads the watched file from disk, parses the file,
+// and updates FileWatcher current data.
+func (fw *FileWatcher) ReadFile() error {
+	data, err := ioutil.ReadFile(fw.fileName)
+	switch {
+	case err != nil:
+		return fmt.Errorf("failed to read file %s: %v", fw.fileName, err)
+	case len(data) == 0:
+		return fmt.Errorf("file %s contains no data", fw.fileName)
+	}
+
+	return nil
+}
+
+// handleEvent filters events and gracefully terminates the operator
+// if a write, remove or create event is received for the watched file.
+func (fw *FileWatcher) handleEvent(event fsnotify.Event) {
+	if !(isWrite(event) || isRemove(event) || isCreate(event)) {
+		return
+	}
+	log.Info("watched file changed", "event", event)
+	os.Exit(1)
```

Comment:
Should we exit 0 given it's a graceful shutdown?

Reply:
err, instead of calling os.Exit, we should be pushing into the stop channel

Reply:
@ironcladlou the stop channel is receive only. I can create a send channel that will trigger a `os.Exit()`.

Reply:
@ironcladlou commit `9d5ed0b` addresses this issue, ptal.

Reply:
https://github.com/openshift/cluster-ingress-operator/blob/master/cmd/ingress-operator/start.go#L126

For example, instead of:

```go
return op.Start(signals.SetupSignalHandler())
```

We could stop on multiple conditions:

```go
stop := make(chan struct{})

// Stop if we receive a signal
go func() {
  stop <- signals.SetupSignalHandler()
}()

// Stop if the cert file changes
go func() {
  // Making up a new function here that blocks until the file changes
  event <- fileWatcher.Watch(cert)
  // Log the event and stop
  stop <- struct{}{}
}()

return op.Start(stop)
```
```